### DISCUSSION TITLE
v1.7.0

### DISCUSSION THREAD

**Discussion by markerikson** [reduxjs/redux-toolkit#1821] [Announcements]
This feature release has a wide variety of API improvements:

- updates RTK Query with support for SSR and rehydration
- allows sharing mutation results across components
- adds a new `currentData` field to query results
- adds several new options for customizing endpoints and base queries
- adds support for async `condition` options in `createAsyncThunk`
- updates `createSlice/createReducer` to accept a "lazy state initializer" function
- updates `createSlice` to avoid potential circular dependency issues by lazy-building its reducer
- updates Reselect and Redux-Thunk to the latest versions with much-improved TS support and new selector customization options
- Fixes a number of small code and types issues

```bash
npm i @reduxjs/toolkit@latest

yarn add @reduxjs/toolkit@latest
```

#### RTK Query SSR and Rehydration Support

RTK Query now has support for SSR scenarios, such as the `getStaticProps/getServerSideProps` APIs in Next.js.  Queries can be executed on the server using the existing `dispatch(someEndpoint.initiate())` thunks, and then collected using the new `await Promise.all(api.getRunningOperationPromises())` method.

API definitions can then provide an `extractRehydrationInfo` method that looks for a specific action type containing the fetched data, and return the data to initialize the API cache section of the store state.

The related `api.util.getRunningOperationPromise()` API adds a building block that may enable future support for React Suspense as well, and we'd encourage users to experiment with this idea.

#### Sharing Mutation Results Across Components

Mutation hooks provide status of in-progress requests, but as originally designed that information was unique per-component - there was no way for another component to see that request status data.  But, we had several requests to enable this use case.  

`useMutation` hooks now support a `fixedCacheKey` option that will store the result status in a common location, so multiple components can read the request status if needed.

This does mean that the data cannot easily be cleaned up automatically, so the mutation status object now includes a `reset()` function that can be used to clear that data.

#### Data Loading Updates

Query results now include a `currentData` field, which contains the latest data cached from the server for the _current_ query arg. Additionally, `transformResponse` now receives the query arg as a parameter.  These can be used to add additional derivation logic in cases when a hooks query arg has changed to represent a different value and the existing data no longer conceptually makes sense to keep displaying.

#### Data Serialization and Base Query Improvements

RTK Query originally only did shallow checks for query arg fields to determine if values had changed.  This caused issues with infinite loops depending on user input.

The query hooks now use a "serialized stable value" hook internally to do more consistent comparisons of query args and eliminate those problems.

Also, `fetchBaseQuery` now supports a `paramsSerializer` option that allows customization of query string generation from the provided arguments, which enables better interaction with some backend APIs.

The `BaseQueryApi` and `prepareheaders` args now include fields for `endpoint` name, `type` to indicate if it's a query or mutation, and `forced` to indicate a re-fetch even if there was already a cache entry.  These can be used to help determine headers like `Cache-Control: no-cache`.

#### Other RTK Query Improvements

API objects now have a `selectInvalidatedBy` function that accepts a root state object and an array of query tag objects, and returns a list of details on endpoints that would be invalidated.  This can be used to help implement optimistic updates of paginated lists.

Fixed an issue serializing a query arg of `undefined`.  Related, an empty JSON body now is stored as `null` instead of `undefined`.

There are now dev warnings for potential mistakes in endpoint setup, like a query function that does not return a `data` field.

Lazy query trigger promises can now be unwrapped similar to mutations.

Fixed a type error that led the endpoint return type to be erroneously used as a state key, which caused generated selectors to have an inferred `state: never` argument.

Fixed `transformResponse` to correctly receive the `originalArgs` as its third parameter.

`api.util.resetApiState` will now clear out cached values in `useQuery` hooks.

The `RetryOptions` interface is now exported, which resolves a TS build error when using the hooks with TS declarations.

#### `createSlice` Lazy Reducers and Circular Dependencies

For the last couple years we've specifically recommended [using a "feature folder" structure with a single "slice" file of logic per feature](https://redux.js.org/style-guide/style-guide#structure-files-as-feature-folders-with-single-file-logic), and `createSlice` makes that pattern really easy - no need to have separate folders and files for `/actions` and `/constants` any more.

The one downside to the "slice file" pattern is in cases when slice A needs to import actions from slice B to respond to them, and slice B also needs to listen to slice A.  This circular import then causes runtime errors, because one of the modules will not have finished initializing by the time the other executes the module body.  That causes the exports to be undefined, and `createSlice` throws an error because you can't pass `undefined` to `builder.addCase()` in `extraReducers`.  (Or, worse, there's no obvious error and things break later.)

There are [well-known patterns for breaking circular dependencies](https://medium.com/visual-development/how-to-fix-nasty-circular-dependency-issues-once-and-for-all-in-javascript-typescript-a04c987cf0de), typically requiring extracting shared logic into a separate file.  For RTK, this usually means [calling `createAction` separately, and importing those action creators into both slices](https://redux-toolkit.js.org/usage/usage-guide#exporting-and-using-slices).

While this is a rarer problem, it's one that _can_ happen in real usage, and it's also been a semi-frequently listed concern from users who didn't want to use RTK.

We've updated `createSlice` to now lazily create its reducer function _the first time you try to call it_.  That delay in instantiation should **eliminate circular dependencies as a runtime error in `createSlice`**.

#### `createAsyncThunk` Improvements

The `condition` option may now be `async`, which enables scenarios like checking if an existing operation is running and resolving the promise when the other instance is done.

If an `idGenerator` function is provided, it will now be given the `thunkArg` value as a parameter, which enables generating custom IDs based on the request data.

The `createAsyncThunk` types were updated to correctly handle type inference when using `rejectWithValue()`.

#### Other RTK Improvements

`createSlice` and `createReducer` now accept a "lazy state initializer" function as the `initialState` argument.  If provided, the initializer will be called to produce a new initial state value any time the reducer is given `undefined` as its state argument.  This can be useful for cases like reading from `localStorage`, as well as testing.

The `isPlainObject` util has been updated to match the implementation in other Redux libs.

The UMD builds of RTK Query now attach as `window.RTKQ` instead of overwriting `window.RTK`.

Fixed an issue with sourcemap loading due to an incorrect filename replacement.

### Dependency Updates

We've updated our deps to the latest versions:

- [**Reselect 4.1.x**](https://github.com/reduxjs/reselect/releases/tag/v4.1.0): Reselect has brand-new customization capabilities for selectors, including configuring cache sizes > 1 and the ability to run equality checks on selector results.  It also now has completely rewritten TS types that do a much better job of inferring arguments and catch previously broken patterns.
- [**Redux Thunk 2.4.0**](https://github.com/reduxjs/redux-thunk/releases/tag/v2.4.0): The thunk middleware also has improved types, as well as an optional "global override" import to modify the type of `Dispatch` everywhere in the app

We've also lowered RTK's peer dependency on React from `^16.14` to `^16.9`, as we just need hooks to be available.

### Other Redux Development Work

The Redux team has also been working on several other updates to the Redux family of libraries.

#### React-Redux v8.0 Beta

We've rewritten React-Redux to add compatibility with the upcoming React 18 release and converted its codebase to TypeScript.  It still supports React 16.8+/17 via a `/compat` entry point.  We'd appreciate further testing from the community so we can confirm it works as expected in real apps before final release.  For details on the changes, see:

- https://github.com/reduxjs/react-redux/releases/tag/v8.0.0-beta.0

#### RTK "Action Listener Middleware" Alpha

We have been working on [a new "action listener middleware"](https://github.com/reduxjs/redux-toolkit/discussions/1648) that we hope to release in an upcoming version of RTK. It's designed to let users write code that runs in response to dispatched actions and state changes, including simple callbacks and moderately complex async workflows.  The current design appears capable of handling many of the use cases that previously required use of the Redux-Saga or Redux-Observable middlewares, but with a smaller bundle size and simpler API.

The listener middleware is still in alpha, but we'd really appreciate more users testing it out and giving us additional feedback to help us finalize the API and make sure it covers the right use cases.

#### RTK Query CodeGen

The [RTK Query OpenAPI codegen tool](https://redux-toolkit.js.org/rtk-query/usage/code-generation) has been rewritten with new options and improved output.

## What's Changed

* fix "isLoading briefly flips back to `true`" #1519 by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1520
* feat(createAsyncThunk): async condition by @thorn0 in https://github.com/reduxjs/redux-toolkit/pull/1496
* add `arg` to `transformResponse` by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1521
* add `currentData` property to hook results. by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1500
* use `useSerializedStableValue` for value comparison by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1533
* fix(useLazyQuery): added docs for preferCache option by @akashshyamdev in https://github.com/reduxjs/redux-toolkit/pull/1541
* correctly handle console logs in tests by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1567
* add `reset` method to useMutation hook by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1476
* allow for "shared component results" using the `useMutation` hook by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1477
* 🐛 Fix bug with `useMutation` shared results by @Shrugsy in https://github.com/reduxjs/redux-toolkit/pull/1616
* pass the ThunkArg to the idGenerator function by @loursbourg in https://github.com/reduxjs/redux-toolkit/pull/1600
* Support a custom paramsSerializer on fetchBaseQuery by @msutkowski in https://github.com/reduxjs/redux-toolkit/pull/1594
* SSR & rehydration support, suspense foundations by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1277
* add `endpoint`, `type` and `forced` to `BaseQueryApi` and `prepareHeaders` by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1656
* split off signature without `AsyncThunkConfig` for better inference by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1644
* Update createReducer to accept a lazy state init function by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/1662
* add `selectInvalidatedBy` by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1665
* Update Yarn from 2.4 to 3.1 by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/1688
* allow for circular references by building reducer lazily on first reducer call by @phryneas in https://github.com/reduxjs/redux-toolkit/pull/1686
* Update deps for 1.7 by @markerikson in https://github.com/reduxjs/redux-toolkit/pull/1692
* fetchBaseQuery: return nullon empty body for JSON. Add DevWarnings by @phryneas in #1699
* Add unwrap to QueryActionCreatorResult and update LazyQueryTrigger by @msutkowski in #1701 
* Only set originalArgs if they're not undefined by @phryneas in #1711
* Treat null as a valid plain object prototype in isPlainObject() in order to sync the util across reduxjs/* repositories by @Ilyklem in #1734
* export RetryOptions interface from retry.ts by @colemars in #1751
* fix: api.util.resetApiState should reset useQuery hooks by @phryneas in #1735
* fix issue where the global RTK object got overwritten in the UMD files by @Antignote in https://github.com/reduxjs/redux-toolkit/pull/1763
* Update dependencies and selector types by @markerikson in #1772
* Fix broken sourcemap output due to bad filename replacement by @markerikson in #1773
* Override unwrap behavior for buildInitiateQuery, update tests by @msutkowski in #1786
* Fix incorrect RTKQ endpoint definition types for correct selector typings by @markerikson in #1818
* Use originalArgs in transformResponse by @msutkowski in #1819

**Full Changelog**: https://github.com/reduxjs/redux-toolkit/compare/v1.6.2...v1.7.0

<hr /><em>This discussion was created from the release <a href='https://github.com/reduxjs/redux-toolkit/releases/tag/v1.7.0'>v1.7.0</a>.</em>

**Comment by paularmstrong** [comment]
> and then collected using the new `await Promise.all(api.getRunningOperationPromises())` method.

Is this a typo (here and in the docs)? It seems that it should be `await Promise.all(api.util.getRunningOperationPromises())`

Also seeing an issue that if you re-render, a new promise replaces the initial promise using a new `requestId` and so the array of pending operations never falls to zero, resulting in the render loop never being able to complete. I'm not sure if this is a bug in rtk-query or an issue with my code, short example:

```ts
// initial render
let html = ReactDOMServer.renderToString(element);
let promises = api.util.getRunningOperationPromises();

while (promises.length > 0) {
  // wait for all API requests
  await Promise.all(promises);
  // re-render after API requests are completed
  html = ReactDOMServer.renderToString(element);
  // check if re-render resulted in more promises to await
  promises = api.util.getRunningOperationPromises();
}
// ensure that no rogue timers are left running.
store.dispatch(api.util.resetApiState());

return html;
```

**Reply by markerikson** [reply]
If it looks like a typo in the release notes, it's almost definitely a typo in the release notes :)  (admittedly there were a _lot_ of release notes to collect here.)

Not sure on the example - @phryneas would have to take a look here.

**Reply by phryneas** [reply]
@paularmstrong I implemented this non-next SSR behaviour solely on decriptions on how people would use it - nobody volunteered to test it or gave any feedback on it during the time we were in beta.  I have never used it myself as I am not doing SSR and would not even know how to set that up. So this is very valuable feedback.

If `ReactDOMServer.renderToString` drops all component-local state (which I am assuming from your example), every component will always still create a new Promise, since on mount of a component, such a promise will be created. It should just instantly resolve in subsequent renders, as opposed to taking the time the request would take on initial render. So if you cannot avoid throwing away the tree every time, it might make sense to check if the promises are already resolved or something like that? 🤔 

Could you do some experimentation there please?

**Reply by paularmstrong** [reply]
@phryneas thank you so much for the work. Totally understand implementing something without much to go on and how hard that is. Sorry I don't have the bandwidth to help out! Just saw the new features and figured it'd be a <10 line change and would try it out.

Here's a case where I expect a total of 3 requests. For this, I expect a total of 3 renders: 

1. renders and starts the request for the user. once resolved:
2. renders, has the user, requests 2 new things, "levelUp" and "accountBenefits". once those are resolved:
3. one last render and no new requests to fill in the data into the rendered tree

After the first render, I see the user request, as expected:
```
[
  Promise {
    <pending>,
    arg: undefined,
    requestId: 'N3jsl_B3rOLOhof0RI13E',
    subscriptionOptions: {
      refetchOnReconnect: undefined,
      refetchOnFocus: undefined,
      pollingInterval: 0
    },
    queryCacheKey: 'user(undefined)',
    abort: [Function: abort],
    unwrap: [Function: unwrap],
    refetch: [Function: refetch],
    unsubscribe: [Function: unsubscribe],
    updateSubscriptionOptions: [Function: updateSubscriptionOptions],
    [Symbol(async_id_symbol)]: 300440,
    [Symbol(trigger_async_id_symbol)]: 300436,
    [Symbol(destroyed)]: { destroyed: false }
  }
]
```

After that resolves, I get the following list of promises, but ALL are pending, including the user request (note new requestId):

```
[
  Promise {
    <pending>,
    arg: undefined,
    requestId: 'f8_AGWKAazgWXmellXx2z',
    subscriptionOptions: {
      refetchOnReconnect: undefined,
      refetchOnFocus: undefined,
      pollingInterval: 0
    },
    queryCacheKey: 'user(undefined)',
    abort: [Function: abort],
    unwrap: [Function: unwrap],
    refetch: [Function: refetch],
    unsubscribe: [Function: unsubscribe],
    updateSubscriptionOptions: [Function: updateSubscriptionOptions],
    [Symbol(async_id_symbol)]: 300564,
    [Symbol(trigger_async_id_symbol)]: 300560,
    [Symbol(destroyed)]: { destroyed: false }
  },
  Promise {
    <pending>,
    arg: { account_pk: 'dc22c72c-3b31-4a97-b4ed-6f2f2a4ffb09' },
    requestId: 'gMFQn72RDvHsRiiy6BCJw',
    subscriptionOptions: {
      refetchOnReconnect: undefined,
      refetchOnFocus: undefined,
      pollingInterval: 0
    },
    queryCacheKey: 'levelUp({"account_pk":"dc22c72c-3b31-4a97-b4ed-6f2f2a4ffb09"})',
    abort: [Function: abort],
    unwrap: [Function: unwrap],
    refetch: [Function: refetch],
    unsubscribe: [Function: unsubscribe],
    updateSubscriptionOptions: [Function: updateSubscriptionOptions],
    [Symbol(async_id_symbol)]: 300624,
    [Symbol(trigger_async_id_symbol)]: 300620,
    [Symbol(destroyed)]: { destroyed: false }
  },
  Promise {
    <pending>,
    arg: { account_pk: 'dc22c72c-3b31-4a97-b4ed-6f2f2a4ffb09' },
    requestId: 'BR6k18X5dn7_qICmhZ5G6',
    subscriptionOptions: {
      refetchOnReconnect: undefined,
      refetchOnFocus: undefined,
      pollingInterval: 0
    },
    queryCacheKey: 'accountBenefits({"account_pk":"dc22c72c-3b31-4a97-b4ed-6f2f2a4ffb09"})',
    abort: [Function: abort],
    unwrap: [Function: unwrap],
    refetch: [Function: refetch],
    unsubscribe: [Function: unsubscribe],
    updateSubscriptionOptions: [Function: updateSubscriptionOptions],
    [Symbol(async_id_symbol)]: 300646,
    [Symbol(trigger_async_id_symbol)]: 300642,
    [Symbol(destroyed)]: { destroyed: false }
  }
]
```

At this point, I will end up in an infinite loop, because each subsequent check, we get 3 promises with new requestIds that will not be resolved immediately. I would like to avoid adding a tick or delay here, but I'm out of ideas given the current implementation.

**Reply by paularmstrong** [reply]
(thinking a little more…)

Adding any sort of tick/timeout/setImmediate to filter the promises after getting them will immediately set the count to zero on a fast system, resulting in the second render never happening.

One other way would be to check if the number of promises is the same on re-render. If so, we can break out of the `while`. My fear is just that it's just not super easy for everyone, nor is it obvious to reason about.

**Reply by paularmstrong** [reply]
This appears to work, implementing the above ^
```js
// render once
let html = ReactDOMServer.renderToString(element);
let promises = api.util.getRunningOperationPromises();
let lastPromisesCount = 0;
while (promises.length > 0 && promises.length !== lastPromisesCount) {
  // wait for all API requests
  await Promise.all(promises);
  lastPromisesCount = promises.length;
  // re-render after API requests are completed
  html = ReactDOMServer.renderToString(element);
  promises = api.util.getRunningOperationPromises();
}
// ensure that no rogue timers are left running.
store.dispatch(api.util.resetApiState());
```

**Reply by phryneas** [reply]
@paularmstrong that snippet looks very useful- could you open a PR to add that to the SSR docs page?
